File: gadt.hs
{-# LANGUAGE GADTs #-}

data Typ a where
    TInt    :: Typ Int
    TString :: Typ String
    TPair   :: (Typ a, Typ b) -> Typ (a, b)

to_string :: Typ a -> a -> String 
to_string TInt x = show x 
to_string TString x = "s:" ++ x 
to_string (TPair (x, y)) (tx, ty) = "(" 
            ++ (to_string x tx) 
            ++ "," 
            ++ (to_string y ty) 
            ++ ")"


main :: IO ()
main = do 
    let s = TPair (TInt, TString)
        t = (1, "a")
    print $ to_string s t 

File: gadt.ml



type _ typ = 
| Int : int typ
| String : string typ
| Pair : 'a typ * 'b typ -> ('a * 'b) typ


let rec to_string : type t. t typ -> t -> string = 
  fun t x -> 
  match t with 
  | Int -> Int.to_string x
  | String -> Printf.sprintf "%S" x
  | Pair (t1, t2) -> 
      let (x1, x2) = x in
      Printf.sprintf "(%s, %s)" (to_string t1 x1) (to_string t2 x2)